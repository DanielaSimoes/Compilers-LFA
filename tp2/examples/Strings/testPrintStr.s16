program testStringDecl () {
    string hello;
    hello = "hello";
    printStr("sou do printStr");
    printStr(hello);
    exit();
}
