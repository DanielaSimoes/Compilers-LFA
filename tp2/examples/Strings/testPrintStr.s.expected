               ; creating a space
               .data
hello:         .space 4

               ; assigning value to hello
               .text
L0:
               ; designating a string value
               .data
str1:          .string "hello"
               .text
               la str1
               store hello

               ; calling print string
               .data
s1:            .string "sou do printStr"
               .text
L2:
               la s1
               jsr printStr

               ; calling print string
               ; checking value of variable
               load hello

               jsr printStr

               ; exiting program
               halt

               ; includes
               halt
printInt:
               ipush 28
printInt1:     dup2
               iushr
               ipush 0xF
               iand
               jsr pNibble
               dup
               ifeq printInt2
               ipush 4
               isub
               goto printInt1
printInt2:     pop
               ret

pNibble:
               ipush 0x30
               iadd
               dup
               ipush 0x39
               isub
               ifle pNibble1
               ipush 0x0a
               iadd
pNibble1:      write
               ret

printStr:      dup
               ipush 0
               baload
               dup
               ifeq printStr_end
               write
               ipush 1
               iadd
               goto printStr

printStr_end:  write
               ret
